agent/cache/watch.go

```diff
@@ -106,148 +106,192 @@ func (c *Cache) notifyBlockingQuery(ctx context.Context, r getOptions, correlati
 
 	for {
 		// Check context hasn't been canceled
-		if ctx.Err() != nil {
+		select {
+		case <-ctx.Done():
 			return
+		default:
 		}
 
-		// Blocking request
-		r.Info.MinIndex = index
-		res, meta, err := c.getWithIndex(ctx, r)
+		index, failures = c.notifyBlockingQueryOnce(ctx, r, correlationID, cb, index, failures)
+	}
+}
+func (c *Cache) notifyBlockingQueryOnce(
+	ctx context.Context,
+	r getOptions,
+	correlationID string,
+	cb Callback,
+	currIndex uint64,
+	currFailures uint,
+) (index uint64, failures uint) {
```

Comment:
nit: Could we avoid the named return values? I think it makes it harder to follow 🙏🏻